tests/Unit/TestHelpers.hpp

```diff
@@ -91,6 +91,8 @@ T serialize_and_deserialize(const T& t) {
 }
 
 /// Test for copy semantics assuming operator== is implement correctly
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wparentheses"
```

Comment:
1) I cannot reproduce this on my machine with GCC 7 and Clang 4 so I cannot see exactly what it is and so unfortunately cannot investigate this myself.

2) Have we tried placing parentheses around the individual arguments of the arguments to `operator==` (I assume that's where the warning is)

3) Can we at least not put warning `push`es around entire functions? I assume it's lines 132 and 134 that are the issue, so we should restrict the `push` to be around those 3 lines.